libraries/AP_HAL_ChibiOS/DSP.cpp

```diff
@@ -0,0 +1,411 @@
+    // 32   -   4us F7,  5us F4
+    // 64   -   7us F7, 13us F4
+    // 128  -  14us F7, 17us F4
+    // 256  -  29us F7, 28us F4
+    // 512  -  55us F7, 93us F4
+    // 1024 - 131us F7
+    // General case for the magnitudes - see https://stackoverflow.com/questions/42299932/dsp-libraries-rfft-strange-results
+    // The frequency of each of those frequency components are given by k*fs/N
+
+    arm_cmplx_mag_squared_f32(&fft->_rfft_data[2], &fft->_freq_bins[1], fft->_bin_count - 1);
+    fft->_freq_bins[0] = sq(fft->_rfft_data[0]);               // DC
+    fft->_freq_bins[fft->_bin_count] = sq(fft->_rfft_data[1]); // Nyquist
+    fft->_rfft_data[fft->_window_size] = fft->_rfft_data[1]; // Nyquist for the interpolator
+    fft->_rfft_data[fft->_window_size + 1] = 0;
+    // find the maximum power in the range we are interested in
+    float max_value = 0;
+    uint16_t bin_range = (end_bin - start_bin) + 1;
+    arm_max_f32(&fft->_freq_bins[start_bin], bin_range, &max_value, &fft->_max_energy_bin);
+    fft->_max_energy_bin += start_bin;
+    // scale the power to account for the input window
+    arm_scale_f32(fft->_freq_bins, fft->_window_scale, fft->_freq_bins, fft->_bin_count);
+
+    TIMER_END(_arm_cmplx_mag_f32_timer);
+    fft->_update_step++;
+}
+    
+// step 6: find the bin with the highest energy and interpolate the required frequency
+uint16_t DSP::step_calc_frequencies(FFTWindowStateARM* fft, uint16_t start_bin, uint16_t end_bin)
+{
+    TIMER_START(_step_calc_frequencies);
+
+    if (is_zero(fft->_freq_bins[fft->_max_energy_bin])) {
+        fft->_max_bin_freq = start_bin * fft->_bin_resolution;
+    }
+    else {
+        // It turns out that Jain is pretty good and works with only magnitudes, but Candan is significantly better
+        // if you have access to the complex values and Quinn is a little better still. Quinn is computationally 
+        // more expensive, but compared to the overall FFT cost seems worth it.
+        fft->_max_bin_freq = (fft->_max_energy_bin + calculate_quinns_second_estimator(fft, fft->_rfft_data, fft->_max_energy_bin)) * fft->_bin_resolution;
+    }
+
+    TIMER_END(_step_calc_frequencies);
+
+#if DEBUG_FFT
+    _output_count++;
+    // outputs at approx 1hz
+    if (_output_count % (400 / fft->_update_steps) == 0) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "FFT: en:%.4f t1:%luus, t2:%luus, t3:%luus, t4:%luus, t5:%luus",
+                        fft->_window_scale, _arm_cfft_f32_timer._timer_avg, _bitreversal_timer._timer_avg, _stage_rfft_f32_timer._timer_avg, _arm_cmplx_mag_f32_timer._timer_avg, _step_calc_frequencies._timer_avg);
+    }
+#endif
+    fft->_update_step++;
+
+    return fft->_max_energy_bin;
+}
+
+// Interpolate center frequency using http://users.metu.edu.tr/ccandan//pub_dir/FineDopplerEst_IEEE_SPL_June2011.pdf
+// This is slightly less accurate than Quinn, but much cheaper to calculate
+float DSP::calculate_candans_estimator(FFTWindowStateARM* fft, uint8_t k_max)
```

Comment:
we could pass FFTWindowStateARM by const ref here

Reply:
Thanks fixed